lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,277 @@
+          post[:cardOnFile][:indicator] = options[:card_on_file][:transaction_id]
+        end
+        if options[:hotel].present?
+          post[:hotel] = {}
+          post[:hotel][:estimatedDays] = options[:hotel][:estimated_days]
+        end
+        if options[:pin].present?
+          post[:pin] = {}
+          post[:pin][:block] = options[:pin][:block]
+          post[:pin][:ksn] = options[:pin][:ksn]
+        end
+      end
+
+      def add_device(post, options)
+        if options[:device].present?
+          post[:device] = {}
+          post[:device][:manufacturer] = options[:device][:manufacturer]
+          post[:device][:model] = options[:device][:model]
+          post[:device][:promptPostalCode] = options[:device][:prompt_postal_code]
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, credit_card, options)
+        post[:card] = {}
+        if options[:token].present?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = options[:token][:value]
+        else
+          if credit_card.respond_to?(:number)
+            post[:card][:expirationDate] = "0#{credit_card.month}#{credit_card.year.to_s[2..4]}"
+            post[:card][:number] = credit_card.number
+            post[:card][:entryMode] = options[:entry_mode]
+            post[:card][:present] = options[:present]
+            post[:card][:securityCode] = {}
+            if options[:security_code].present?
+              post[:card][:securityCode][:indicator] = options[:security_code][:indicator]
+              post[:card][:securityCode][:value] = options[:security_code][:value]
+            end
+          end
+          end
+
+      end
+
+      def add_customer(post, options)
+        if address = options[:customer]
+          post[:customer] = {}
+          post[:customer][:addressLine1] = address[:address_line1] if address[:address_line1]
+          post[:customer][:firstName] = address[:first_name]
+          post[:customer][:lastName] = address[:last_name]
+          post[:customer][:postalCode] = address[:postal_code]
+        end
+      end
+
+      def commit(action, parameters, headers)
+        url_postfix = action == 'accesstoken' ? 'credentials' : 'transactions'
+        url = (test? ? "#{test_url}#{url_postfix}/#{action}" : "#{live_url}#{url_postfix}/#{action}")
```

Comment:
I have updated your commit method as per the standards. please rectify requests for failing test cases (test_failed_authorize, test_failed_capture etc.)